src/u2fprotocol.rs

```diff
@@ -322,135 +214,91 @@ fn sendrecv<T>(dev: &mut T, cmd: u8, send: &[u8]) -> io::Result<Vec<u8>>
-            init_sent = true;
+    while data.len() > 0 {
+        // HID_RPT_SIZE+1 since we need to prefix this with a record index.
+        let mut frame = [0; HID_RPT_SIZE + 1];
+
+        if data.len() == send.len() {
+            let max = cmp::min(data.len(), INIT_DATA_SIZE);
+            let uf = U2FHIDInit::new(dev, cmd, send.len(), &data[..max]);
+            uf.to_bytes(&mut frame[1..]);
+            data = &data[max..];
         } else {
-            let mut uf = U2FHIDCont {
-                cid: dev.get_cid(),
-                seq: sequence,
-                data: [0; CONT_DATA_SIZE],
-            };
-            set_data(&mut uf.data, &mut data_itr, CONT_DATA_SIZE);
+            // First cont package has seq=1.
+            let max = cmp::min(data.len(), CONT_DATA_SIZE);
+            let uf = U2FHIDCont::new(dev, sequence, &data[..max]);
+            uf.to_bytes(&mut frame[1..]);
+            data = &data[max..];
             sequence += 1;
-            frame[1..].clone_from_slice(to_u8_array(&uf));
         }
 
         if log_enabled!(log::LogLevel::Trace) {
             let parts: Vec<String> = frame.iter().map(|byte| format!("{:02x}", byte)).collect();
             trace!("USB send: {}", parts.join(""));
         }
 
-        if let Err(er) = dev.write(&frame) {
-            return Err(er);
-        };
+        try!(dev.write(&frame));
     }
 
     // Now we read. This happens in 2 chunks: The initial packet, which has the
     // size we expect overall, then continuation packets, which will fill in
     // data until we have everything.
-    let mut frame: [u8; HID_RPT_SIZE] = [0u8; HID_RPT_SIZE];
-    let datalen: usize;
-    let mut data: Vec<u8>;
+    let mut frame = [0u8; HID_RPT_SIZE];
 
     // TODO Check the status of the read, figure out how we'll deal with timeouts.
     dev.read(&mut frame)?;
     let mut recvlen = INIT_DATA_SIZE;
 
     // We'll get an init packet back from USB, open it to see how much we'll be
-    // reading overall. (should unpack to an init struct). Scope this to deal
-    // with the lifetime of the frame borrow in from_u8_array.
-    {
-        let info_frame: &U2FHIDInit = from_u8_array(&frame);
-
-        // Read until we've exhausted the total read amount or error out
-        datalen = (info_frame.bcnth as usize) << 8 | (info_frame.bcntl as usize);
-        data = Vec::with_capacity(datalen);
-
-        let clone_len: usize;
-        if datalen < recvlen {
-            clone_len = datalen;
-        } else {
-            clone_len = recvlen;
-        }
-        data.extend(info_frame.data[0..clone_len].iter().cloned());
-    }
-    sequence = 0;
+    // reading overall. (should unpack to an init struct).
+    let info_frame = U2FHIDInit::from_bytes(&frame)?;
+
+    // Read until we've exhausted the total read amount or error out
+    let datalen = info_frame.bcnt();
+    let mut data = Vec::with_capacity(datalen);
+    let clone_len = cmp::min(datalen, recvlen);
```

Comment:
Nit: Since recvlen is still constant here, this call to `min` feels a little stylistically weird, like "wait, recvlen hasn't been updated yet, is that OK?"

I would have this min be against the `INIT_DATA_SIZE` to be clearer, and declare recvlen before entering the loop.